habitat_baselines/il/data/edfe_data.py

```diff
@@ -0,0 +1,224 @@
+                self.scene_episode_dict[ep.scene_id] = [ep.episode_id]
+            else:
+                self.scene_episode_dict[ep.scene_id].append(ep.episode_id)
+
+        self.disk_cache_exists = self.check_cache_exists()
+
+        if not self.disk_cache_exists:
+            try:
+                self.make_dataset_dirs()
+            except FileExistsError:
+                pass
+            """
+            for each scene > load scene in memory > save frames for each
+            episode corresponding to each scene
+            """
+            print("Saving rgb, seg, depth frames to disk.")
+
+            for scene in tqdm(list(self.scene_episode_dict.keys())):
+
+                self.load_scene(scene)
+
+                for ep_id in tqdm(self.scene_episode_dict[scene]):
+
+                    episode = next(
+                        ep for ep in self.episodes if ep.episode_id == ep_id
+                    )
+
+                    pos_queue = episode.shortest_paths[0]
+                    random_pos = random.sample(pos_queue, 9)
+                    self.save_frames(random_pos, ep_id)
+
+            print("Saved all episodes' frames to disk. Frame dataset ready.")
+        self.env.close()
+
+        self.rgb_list = sorted(
+            os.listdir(self.dataset_path.format(split=self.mode, type="rgb"))
+        )
+        self.depth_list = sorted(
+            os.listdir(self.dataset_path.format(split=self.mode, type="depth"))
+        )
+        self.seg_list = sorted(
+            os.listdir(self.dataset_path.format(split=self.mode, type="seg"))
+        )
+
+    def save_frames(
+        self, pos_queue: List[ShortestPathPoint], episode_id
+    ) -> None:
+        r"""Writes rgb, seg, depth frames to disk.
+        """
+
+        for idx, pos in enumerate(pos_queue):
+
+            observation = self.env.sim.get_observations_at(
+                pos.position, pos.rotation
+            )
+
+            depth = observation["depth"]
+            rgb = observation["rgb"]
+            seg = observation["semantic"]
+            seg = seg % 40
+            seg = seg.astype(np.uint8)
+
+            if random.random() < 0.8:
+                split = "train"
+            else:
+                split = "val"
+
+            rgb_frame_path = os.path.join(
+                self.dataset_path.format(split=split, type="rgb"),
+                "ep_" + episode_id + "_{0:0=3d}".format(idx),
+            )
+            depth_frame_path = os.path.join(
+                self.dataset_path.format(split=split, type="depth"),
+                "ep_" + episode_id + "_{0:0=3d}".format(idx),
+            )
+            seg_frame_path = os.path.join(
+                self.dataset_path.format(split=split, type="seg"),
+                "ep_" + episode_id + "_{0:0=3d}".format(idx),
+            )
+
+            cv2.imwrite(rgb_frame_path + ".jpg", rgb)
+            cv2.imwrite(depth_frame_path + ".jpg", depth * 255)
+            np.savez_compressed(seg_frame_path, seg)
```

Comment:
I highly recommending putting these into a database like https://www.h5py.org/ or https://lmdb.readthedocs.io/en/release/

Reading out of databases like this is considerably faster than reading from thousands of files on disk.

Reply:
Alright, I'll make the changes. Thank you

Reply:
@erikwijmans 

There are a total of 360k images (100k * 3 train + 20k * 3 test) that need to be saved, before training starts. Currently, the dataset takes about 5.1 GB disk space (4.1 GB train + 1 GB test). 

Even though reading out files from an h5 database will be fast, saving the whole dataset into train.h5 and test.h5 will likely have the below problems - 
 
1. Take much more disk space (even gzip compressed ones)
2. While writing to H5 format - we can't load all the (360k) frames into memory before writing them out to the h5 file , and appending images to the h5 database, episode by episode will be quite memory expensive too.

Could you help me with what can be done? Thank you

Reply:
You can use PIL + LMDB to keep things compressed.  LDMB is a key-value storage and will just store arbitrary binary. I know you can use PIL to write to/read from https://docs.python.org/3/library/io.html#io.BytesIO (cv2 may also be able to do this, but you need to be careful with rgb vs. bgr).  jpeg is pretty slow to decode however, so storing uncompressed is common.

For h5, you can pre-allocate everything and just write directly into the h5 file for each image.  No need to sore everything in memory and write all at once.

Reply:
@erikwijmans 

Thank you for the explanation. I have made the suggested changes in the latest commit-

- fix for semantic segmentation annotations
- LMDB database added

If the LMDB implementation looks good, I'll make the same changes for the VQA, NAV trainer datasets.

Reply:
@erikwijmans I was implementing the LMDB dataset for the VQA trainer. I noticed that storing uncompressed numpy arrrays will take about 10x more storage. More storage was okay for the smaller EDFE dataset. But the dataset required for VQA and NAV trainers is huge (avg 95 images per episode, for ~13k episodes = ~1250k images). The normal jpeg dataset saved on disk occupied ~37 GB disk space. The LMDB dataset, however, has taken up ~100 GB for storing episode frames from just 20 scenes (out of 67 train+test scenes). So it is likely to take >300 GB storage all in all.

Just for your information - with the basic file system (jpeg encoded images), the network took about 4 minutes per epoch on a batch size of 50. I think that's pretty reasonable.

Please share your thoughts on the same. Thank you

Reply:
@erikwijmans Could you please share your thoughts on the previous comment?

Reply:
Storing things as jpeg in a database is fine, but storing things as individual images is really inefficient, especially if the access will be done via a networked filesystem as is the case on the HPC clusters that are common in DL.   An option going forward will be this: https://github.com/tmbdev/webdataset as it is also likely getting integrated into pytorch: https://github.com/pytorch/pytorch/issues/38419 and is easier to work with than lmdb.